File: optimization-greedy-algorithms-problems-pro/src/test/java/com/greedy/problems/JobSequencingTest.java
package com.greedy.problems;

import com.greedy.utils.DataStructures.Job;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;

/**
 * Unit tests for the {@link JobSequencing} class.
 * Tests various scenarios to ensure the greedy algorithm correctly maximizes profit
 * while adhering to job deadlines.
 */
public class JobSequencingTest {

    private JobSequencing solver;

    @BeforeEach
    void setUp() {
        solver = new JobSequencing();
    }

    @Test
    void testEmptyJobs() {
        Job[] jobs = {};
        assertEquals(0, solver.findMaxProfit(jobs), "Should return 0 profit for empty jobs array.");
    }

    @Test
    void testNullJobs() {
        Job[] jobs = null;
        assertEquals(0, solver.findMaxProfit(jobs), "Should return 0 profit for null jobs array.");
    }

    @Test
    void testSingleJob() {
        Job[] jobs = {new Job('a', 1, 100)};
        assertEquals(100, solver.findMaxProfit(jobs), "Single job should be taken if valid.");
    }

    @Test
    void testMultipleJobs_AllCanBeDone() {
        Job[] jobs = {
                new Job('a', 2, 30),
                new Job('b', 1, 20),
                new Job('c', 3, 10)
        };
        // Sorted by profit: a(30), b(20), c(10)
        // a (d=2): slot 1
        // b (d=1): slot 0
        // c (d=3): slot 2
        // Profit: 30+20+10 = 60
        assertEquals(60, solver.findMaxProfit(jobs), "All jobs should be taken if compatible.");
    }

    @Test
    void testStandardCase_Example1() {
        Job[] jobs = {
                new Job('a', 4, 20),
                new Job('b', 1, 10),
                new Job('c', 1, 40),
                new Job('d', 1, 30)
        };
        // Sorted by profit: c(40), d(30), a(20), b(10)
        // Max deadline = 4. Slots: [_,_,_,_]
        // 1. Job 'c' (d=1, p=40): slot 0. Schedule: [c,_,_,_]. Profit = 40.
        // 2. Job 'd' (d=1, p=30): slot 0 taken. No slot <=0. Skip.
        // 3. Job 'a' (d=4, p=20): slot 3. Schedule: [c,_,_,a]. Profit = 40+20=60.
        // 4. Job 'b' (d=1, p=10): slot 0 taken. No slot <=0. Skip.
        // Max Profit: 60.
        assertEquals(60, solver.findMaxProfit(jobs), "Should correctly sequence jobs for Example 1.");
    }

    @Test
    void testStandardCase_Example2() {
        Job[] jobs = {
                new Job('a', 2, 100),
                new Job('b', 1, 19),
                new Job('c', 2, 27),
                new Job('d', 1, 25),
                new Job('e', 3, 15)
        };
        // Sorted by profit: a(100), c(27), d(25), e(15), b(19)
        // Max deadline = 3. Slots: [_,_,_]
        // 1. Job 'a' (d=2, p=100): slot 1. Schedule: [_,a,_]. Profit = 100.
        // 2. Job 'c' (d=2, p=27): slot 0. Schedule: [c,a,_]. Profit = 100+27=127.
        // 3. Job 'd' (d=1, p=25): slot 0 taken. No slot <=0. Skip.
        // 4. Job 'e' (d=3, p=15): slot 2. Schedule: [c,a,e]. Profit = 127+15=142.
        // 5. Job 'b' (d=1, p=19): slot 0 taken. No slot <=0. Skip.
        // Max Profit: 142.
        assertEquals(142, solver.findMaxProfit(jobs), "Should correctly sequence jobs for Example 2.");
    }

    @Test
    void testJobsWithSameProfit() {
        Job[] jobs = {
                new Job('a', 2, 50),
                new Job('b', 1, 50),
                new Job('c', 2, 30)
        };
        // Sorted by profit (desc): a(50), b(50), c(30).
        // If tie-breaking is by deadline, then b before a. But actual sort is just profit.
        // Assume default sort on char ID: a, b, c. Or just order in array.
        // Max deadline = 2. Slots: [_,_]
        // Case 1: If 'a' processed first among 50-profit jobs.
        // 1. Job 'a' (d=2, p=50): slot 1. Schedule: [_,a]. Profit = 50.
        // 2. Job 'b' (d=1, p=50): slot 0. Schedule: [b,a]. Profit = 50+50=100.
        // 3. Job 'c' (d=2, p=30): slot 0 taken, slot 1 taken. Skip.
        // Max Profit: 100.
        //
        // Case 2: If 'b' processed first among 50-profit jobs.
        // 1. Job 'b' (d=1, p=50): slot 0. Schedule: [b,_]. Profit = 50.
        // 2. Job 'a' (d=2, p=50): slot 1. Schedule: [b,a]. Profit = 50+50=100.
        // 3. Job 'c' (d=2, p=30): slot 0 taken, slot 1 taken. Skip.
        // Max Profit: 100.
        assertEquals(100, solver.findMaxProfit(jobs), "Should handle jobs with same profit correctly.");
    }

    @Test
    void testJobsWithNoFeasibleSchedule() {
        Job[] jobs = {
                new Job('a', 1, 100),
                new Job('b', 1, 50)
        };
        // Sorted by profit: a(100), b(50)
        // Max deadline = 1. Slots: [_]
        // 1. Job 'a' (d=1, p=100): slot 0. Schedule: [a]. Profit = 100.
        // 2. Job 'b' (d=1, p=50): slot 0 taken. Skip.
        // Max Profit: 100.
        assertEquals(100, solver.findMaxProfit(jobs), "Should select only one job if others are infeasible due to deadlines.");
    }

    @Test
    void testJobsWithDeadlineGreaterThanTotalSlots() {
        Job[] jobs = {
                new Job('a', 10, 100),
                new Job('b', 1, 50)
        };
        // Sorted by profit: a(100), b(50)
        // Max deadline = 10. Slots array of size 10.
        // 1. Job 'a' (d=10, p=100): slot 9 (max available <= deadline).
        // 2. Job 'b' (d=1, p=50): slot 0.
        // Profit: 100 + 50 = 150.
        assertEquals(150, solver.findMaxProfit(jobs), "Should handle deadlines larger than number of jobs.");
    }

    @Test
    void testJobsWithZeroProfit() {
        Job[] jobs = {
                new Job('a', 2, 50),
                new Job('b', 1, 0),
                new Job('c', 2, 30)
        };
        // Sorted by profit: a(50), c(30), b(0)
        // Max deadline = 2. Slots: [_,_]
        // 1. Job 'a' (d=2, p=50): slot 1. Schedule: [_,a]. Profit = 50.
        // 2. Job 'c' (d=2, p=30): slot 0. Schedule: [c,a]. Profit = 50+30=80.
        // 3. Job 'b' (d=1, p=0): slot 0 taken. Skip.
        // Max Profit: 80.
        assertEquals(80, solver.findMaxProfit(jobs), "Should ignore zero profit jobs if better options exist.");
    }
}

File: leetcode-greedy-algorithms-solutions-advanced/tests/jobSequencing.test.js
```javascript
/**
 * @fileoverview Test suite for the Job Sequencing with Deadlines Problem.
 */

import jobSequencing from '../src/problems/jobSequencing';

describe('jobSequencing', () => {
  // Test Case 1: Basic scenario
  test('should correctly sequence jobs for a basic scenario', () => {
    const jobs = [
      { id: 'a', deadline: 2, profit: 100 },
      { id: 'b', deadline: 1, profit: 19 },
      { id: 'c', deadline: 2, profit: 27 },
      { id: 'd', deadline: 1, profit: 25 },
      { id: 'e', deadline: 3, profit: 15 }
    ];
    // Sorted by profit (desc): a(100), c(27), d(25), b(19), e(15)
    // Max deadline = 3. Slots: [null, null, null]
    // 1. Job 'a' (profit 100, deadline 2): Try slot 2, free. slots=['a', 'a', null]. totalProfit=100.
    //    Actually, we schedule into a slot, so slots[0] is time 1, slots[1] is time 2.
    //    Try slot deadline-1 (index 1), free. slots[1] = 'a'. So ['a', null] (max deadline is 2)
    //    No, max deadline is 3. So slots: [null, null, null]
    //    Job 'a' (deadline 2): Try slot 2 (index 1). slots: [null, 'a', null]. totalProfit=100.
    // 2. Job 'c' (profit 27, deadline 2): Try slot 2 (index 1), taken by 'a'. Try slot 1 (index 0), free. slots: ['c', 'a', null]. totalProfit=100+27=127.
    // 3. Job 'd' (profit 25, deadline 1): Try slot 1 (index 0), taken by 'c'. Cannot schedule.
    // 4. Job 'b' (profit 19, deadline 1): Try slot 1 (index 0), taken by 'c'. Cannot schedule.
    // 5. Job 'e' (profit 15, deadline 3): Try slot 3 (index 2), free. slots: ['c', 'a', 'e']. totalProfit=127+15=142.
    const expected = {
      totalProfit: 142,
      sequencedJobs: ['c', 'a', 'e'] // Order reflects chronological execution by slot
    };
    const result = jobSequencing(jobs);
    expect(result.totalProfit).toBe(expected.totalProfit);
    expect(result.sequencedJobs).toEqual(expected.sequencedJobs);
  });

  // Test Case 2: No jobs
  test('should return 0 profit and empty array for no jobs', () => {
    const jobs = [];
    const expected = { totalProfit: 0, sequencedJobs: [] };
    expect(jobSequencing(jobs)).toEqual(expected);
  });

  // Test Case 3: Single job
  test('should schedule a single job with its profit', () => {
    const jobs = [{ id: 'x', deadline: 1, profit: 50 }];
    const expected = { totalProfit: 50, sequencedJobs: ['x'] };
    expect(jobSequencing(jobs)).toEqual(expected);
  });

  // Test Case 4: All jobs have deadline 1
  test('should schedule only the highest profit job if all have deadline 1', () => {
    const jobs = [
      { id: 'a', deadline: 1, profit: 10 },
      { id: 'b', deadline: 1, profit: 20 },
      { id: 'c', deadline: 1, profit: 5 }
    ];
    // Sorted: b(20), a(10), c(5)
    // Only 'b' can be scheduled.
    const expected = { totalProfit: 20, sequencedJobs: ['b'] };
    expect(jobSequencing(jobs)).toEqual(expected);
  });

  // Test Case 5: Jobs with same profit but different deadlines
  test('should prioritize jobs with earlier deadlines if profits are equal (not strictly part of greedy, but good for tie-breaking)', () => {
    const jobs = [
      { id: 'a', deadline: 2, profit: 50 },
      { id: 'b', deadline: 1, profit: 50 }
    ];
    // Sorted by profit: a(50), b(50) -> if stable sort, order depends on original.
    // Let's assume input order is kept if profits are equal.
    // Test for a stable sort that places `a` first among equals:
    const jobsAFirst = [
      { id: 'a', deadline: 2, profit: 50 },
      { id: 'b', deadline: 1, profit: 50 }
    ];
    // 1. Job 'a' (profit 50, deadline 2): slots: [null, 'a']. profit 50.
    // 2. Job 'b' (profit 50, deadline 1): deadline 1, try slot 1 (index 0). free. slots: ['b', 'a']. profit 50+50=100.
    expect(jobSequencing(jobsAFirst)).toEqual({ totalProfit: 100, sequencedJobs: ['b', 'a'] });

    // Test for `b` first among equals:
    const jobsBFirst = [
      { id: 'b', deadline: 1, profit: 50 },
      { id: 'a', deadline: 2, profit: 50 }
    ];
    // 1. Job 'b' (profit 50, deadline 1): slots: ['b', null]. profit 50.
    // 2. Job 'a' (profit 50, deadline 2): deadline 2, try slot 2 (index 1). free. slots: ['b', 'a']. profit 50+50=100.
    expect(jobSequencing(jobsBFirst)).toEqual({ totalProfit: 100, sequencedJobs: ['b', 'a'] });
    // In both cases, the optimal schedule (b, a) is chosen. The tie-breaking rule for same profit
    // doesn't change the outcome, as long as it correctly finds a free slot.
  });

  // Test Case 6: Jobs with deadlines far in the future, but limited number of actual jobs
  test('should handle jobs with distant deadlines efficiently', () => {
    const jobs = [
      { id: 'a', deadline: 100, profit: 10 },
      { id: 'b', deadline: 10, profit: 20 },
      { id: 'c', deadline: 5, profit: 30 }
    ];
    // Sorted by profit: c(30), b(20), a(10)
    // Max deadline = 100.
    // 1. Job 'c' (profit 30, deadline 5): slots[4] = 'c'. totalProfit=30.
    // 2. Job 'b' (profit 20, deadline 10): slots[9] = 'b'. totalProfit=30+20=50.
    // 3. Job 'a' (profit 10, deadline 100): slots[99] = 'a'. totalProfit=50+10=60.
    // Note: The `slots` array will be large, but only few elements are filled.
    const expected = { totalProfit: 60, sequencedJobs: ['c', 'b', 'a'] }; // Chronological
    const result = jobSequencing(jobs);
    expect(result.totalProfit).toBe(expected.totalProfit);
    expect(result.sequencedJobs).toEqual(expected.sequencedJobs);
  });

  // Test Case 7: Complex scenario with many overlapping deadlines and profits
  test('should handle complex job sequencing scenarios', () => {
    const jobs = [
      { id: 'j1', deadline: 4, profit: 20 },
      { id: 'j2', deadline: 1, profit: 10 },
      { id: 'j3', deadline: 1, profit: 40 },
      { id: 'j4', deadline: 1, profit: 30 },
      { id: 'j5', deadline: 3, profit: 35 },
      { id: 'j6', deadline: 2, profit: 25 }
    ];
    // Sorted by profit (desc): j3(40), j5(35), j4(30), j6(25), j1(20), j2(10)
    // Max deadline = 4. Slots: [null, null, null, null] (for times 1, 2, 3, 4)
    // 1. j3 (P40, D1): slots[0]='j3'. P=40.
    // 2. j5 (P35, D3): slots[2]='j5'. P=40+35=75.
    // 3. j4 (P30, D1): Try slot 0, taken. Cannot schedule.
    // 4. j6 (P25, D2): slots[1]='j6'. P=75+25=100.
    // 5. j1 (P20, D4): slots[3]='j1'. P=100+20=120.
    // 6. j2 (P10, D1): Try slot 0, taken. Cannot schedule.
    const expected = { totalProfit: 120, sequencedJobs: ['j3', 'j6', 'j5', 'j1'] };
    const result = jobSequencing(jobs);
    expect(result.totalProfit).toBe(expected.totalProfit);
    expect(result.sequencedJobs).toEqual(expected.sequencedJobs);
  });

  // Test Case 8: Deadlines are all the same, different profits
  test('should pick highest profit job if all deadlines are same but greater than 1', () => {
    const jobs = [
      { id: 'A', deadline: 2, profit: 10 },
      { id: 'B', deadline: 2, profit: 30 },
      { id: 'C', deadline: 2, profit: 20 }
    ];
    // Sorted by profit: B(30), C(20), A(10)
    // Max deadline = 2. Slots: [null, null]
    // 1. B (P30, D2): slots[1]='B'. P=30.
    // 2. C (P20, D2): Try slot 1, taken. Try slot 0, free. slots[0]='C'. P=30+20=50.
    // 3. A (P10, D2): Try slot 1, taken. Try slot 0, taken. Cannot schedule.
    const expected = { totalProfit: 50, sequencedJobs: ['C', 'B'] };
    const result = jobSequencing(jobs);
    expect(result.totalProfit).toBe(expected.totalProfit);
    expect(result.sequencedJobs).toEqual(expected.sequencedJobs);
  });
});
```